tensorboard/notebook.py

```diff
@@ -0,0 +1,336 @@
+      Shell metacharacters are not supported: e.g., "--logdir 2>&1" will
+      point the logdir at the literal directory named "2>&1".
+  """
+  context = _get_context()
+  try:
+    import IPython
+    import IPython.display
+  except ImportError:
+    IPython = None
+
+  if context == _CONTEXT_NONE:
+    handle = None
+    print("Launching TensorBoard...")
+  else:
+    handle = IPython.display.display(
+        IPython.display.Pretty("Launching TensorBoard..."),
+        display_id=True,
+    )
+  def print_or_update(message):
+    if handle is None:
+      print(message)
+    else:
+      handle.update(IPython.display.Pretty(message))
+
+  parsed_args = shlex.split(args_string, comments=True, posix=True)
+  start_result = manager.start(parsed_args)
+
+  if isinstance(start_result, manager.StartLaunched):
+    _display(
+        port=start_result.info.port,
+        print_message=False,
+        display_handle=handle,
+    )
+
+  elif isinstance(start_result, manager.StartReused):
+    template = (
+        "Reusing TensorBoard on port {port} (pid {pid}), started {delta} ago. "
+        "(Use '!kill {pid}' to kill it.)"
+    )
+    message = template.format(
+        port=start_result.info.port,
+        pid=start_result.info.pid,
+        delta=_time_delta_from_info(start_result.info),
+    )
+    print_or_update(message)
+    _display(
+        port=start_result.info.port,
+        print_message=False,
+        display_handle=None,
+    )
+
+  elif isinstance(start_result, manager.StartFailed):
+    def format_stream(name, value):
+      if value == "":
+        return ""
+      elif value is None:
+        return "\n<could not read %s>" % name
+      else:
+        return "\nContents of %s:\n%s" % (name, value.strip())
+    message = (
+        "ERROR: Failed to launch TensorBoard (exited with %d).%s%s" %
+        (
+            start_result.exit_code,
+            format_stream("stderr", start_result.stderr),
+            format_stream("stdout", start_result.stdout),
+        )
+    )
+    print_or_update(message)
+
+  elif isinstance(start_result, manager.StartTimedOut):
+    message = (
+        "ERROR: Timed out waiting for TensorBoard to start. "
+        "It may still be running as pid %d."
+        % start_result.pid
+    )
+    print_or_update(message)
+
+  else:
+    raise TypeError(
+        "Unexpected result from `manager.start`: %r.\n"
+        "This is a TensorBoard bug; please report it."
+        % start_result
+    )
+
+
+def _time_delta_from_info(info):
+  """Format the elapsed time for the given TensorboardInfo.
+
+  Args:
+    info: A TensorboardInfo value.
+
+  Returns:
+    A human-readable string describing the time since the server
+    described by `info` started: e.g., "2 days, 0:48:58".
+  """
+  now = datetime.datetime.now()
+  then = info.start_time
+  return str(now.replace(microsecond=0) - then.replace(microsecond=0))
+
+
+def _data_source_from_info(info):
```

Comment:
Why does this implementation detail need to be exposed here? Can't it be hidden inside TensorboardInfo or manager?

Also -- just noticed that 'b' in TensorboardInfo is not consistent with our general CamelCase usage of "TensorBoard".

Reply:
> Why does this implementation detail need to be exposed here?

Exposed where? The `_data_source_from_info` function is private to this
module.

> Can't it be hidden inside TensorboardInfo or manager?

Per Nick’s comments on the (Google-internal) design doc, we’re trying to
keep the `manager` interface minimal, and implement helpers like this
and the get-info-by-port logic in `notebook` for now.

> Also -- just noticed that 'b' in TensorboardInfo is not consistent
> with our general CamelCase usage of "TensorBoard".

This is by design. The camel-case “TensorBoard” is the human-readable
name. Identifiers like `TensorboardInfo` and `GithubPlugin` and
`UserIdGenerator` are (in my experience) best capitalized thus, because
otherwise their lower-camel-case forms `tensorBoardInfo` and
`gitHubPlugin` and `userIDGenerator` are awkward.

Reply:
Per our off-line conversation, rename to TensorBoardInfo for consistency with naming across the rest of the codebase.

Reply:
Also, move _data_source_from_info() to inside manager.py to avoid notebook worrying about the details of the data source here.

Reply:
Moved the new function `data_source_from_info`; will rename the existing
`TensorboardInfo` separately.